File: formacOS/Tests/AnyClipCoreTests/VersionNegotiatorTests.swift
import Testing
@testable import AnyClipCore

private func v(_ major: Int, _ minor: Int, app: String = "1.0.0") -> VersionInfo {
    VersionInfo(appVersion: app, protocolMajor: major, protocolMinor: minor)
}

@Test func sameVersionIsCompatible() {
    #expect(negotiate(local: v(1, 0), peer: v(1, 0)) == .compatible)
}

@Test func peerOlderMinorLinksWithHint() {
    let r = negotiate(local: v(1, 2), peer: v(1, 0))
    #expect(r == .peerOlderMinor)
    #expect(linkAllowed(r))
}

@Test func peerNewerMinorLinksWithHint() {
    let r = negotiate(local: v(1, 0), peer: v(1, 2))
    #expect(r == .peerNewerMinor)
    #expect(linkAllowed(r))
}

@Test func peerOlderMajorRefused() {
    let r = negotiate(local: v(2, 0), peer: v(1, 5))
    #expect(r == .peerOlderMajor)
    #expect(!linkAllowed(r))
}

@Test func peerNewerMajorRefused() {
    let r = negotiate(local: v(1, 9), peer: v(2, 0))
    #expect(r == .peerNewerMajor)
    #expect(!linkAllowed(r))
}

@Test func appVersionNeverAffectsOutcome() {
    #expect(negotiate(local: v(1, 0, app: "1.0.0"), peer: v(1, 0, app: "9.9.9")) == .compatible)
}

@Test func rawValuesMatchPythonEnum() {
    #expect(Compatibility.compatible.rawValue == "compatible")
    #expect(Compatibility.peerOlderMinor.rawValue == "peer_older_minor")
    #expect(Compatibility.peerNewerMinor.rawValue == "peer_newer_minor")
    #expect(Compatibility.peerOlderMajor.rawValue == "peer_older_major")
    #expect(Compatibility.peerNewerMajor.rawValue == "peer_newer_major")
}

File: forwindows/tests/AnyClipCore.Tests/VersionNegotiatorTests.cs
using AnyClip.Core;
using Xunit;

namespace AnyClip.Core.Tests;

public class VersionNegotiatorTests
{
    private static VersionInfo V(int major, int minor, string app = "1.0.0") =>
        new(app, major, minor);

    [Fact] public void Same() =>
        Assert.Equal(Compatibility.Compatible, VersionNegotiator.Negotiate(V(1, 0), V(1, 0)));

    [Fact]
    public void PeerOlderMinorLinks()
    {
        var r = VersionNegotiator.Negotiate(V(1, 2), V(1, 0));
        Assert.Equal(Compatibility.PeerOlderMinor, r);
        Assert.True(VersionNegotiator.LinkAllowed(r));
    }

    [Fact]
    public void PeerNewerMinorLinks()
    {
        var r = VersionNegotiator.Negotiate(V(1, 0), V(1, 2));
        Assert.Equal(Compatibility.PeerNewerMinor, r);
        Assert.True(VersionNegotiator.LinkAllowed(r));
    }

    [Fact]
    public void MajorMismatchRefused()
    {
        Assert.False(VersionNegotiator.LinkAllowed(
            VersionNegotiator.Negotiate(V(2, 0), V(1, 5))));
        Assert.False(VersionNegotiator.LinkAllowed(
            VersionNegotiator.Negotiate(V(1, 9), V(2, 0))));
    }

    [Fact]
    public void WireValuesMatchPython()
    {
        Assert.Equal("compatible", VersionNegotiator.WireValue(Compatibility.Compatible));
        Assert.Equal("peer_older_minor", VersionNegotiator.WireValue(Compatibility.PeerOlderMinor));
        Assert.Equal("peer_newer_minor", VersionNegotiator.WireValue(Compatibility.PeerNewerMinor));
        Assert.Equal("peer_older_major", VersionNegotiator.WireValue(Compatibility.PeerOlderMajor));
        Assert.Equal("peer_newer_major", VersionNegotiator.WireValue(Compatibility.PeerNewerMajor));
    }
}
